File: 2021/4/part-one.py
import sys


def find_bingo(board):
    rows = [0] * 5
    cols = [0] * 5
    for i in range(0, 5):
        for j in range(0, 5):
            if board[i][j][1]:
                rows[i] += 1
                cols[i] += 1
                if 5 in rows or 5 in cols:
                    return True
    return False


if __name__ == "__main__":
    with open(sys.argv[1], encoding="utf-8") as f:
        groups = f.read().strip().split("\n\n")

    numbers, groups = groups[0].split(","), groups[1:]
    numbers = [int(x) for x in numbers]

    boards = []
    for g in groups:
        rows = []
        for r in g.replace("  ", " ").split("\n"):
            values = []
            for v in r.strip().split(" "):

                values.append([int(v), False])
            rows.append(values)
        boards.append(rows)

    for n in numbers:
        for i in range(len(boards)):
            for j in range(len(boards[i])):
                boards[i][j] = [[x[0], True] if x[0] == n else x for x in boards[i][j]]
            if find_bingo(boards[i]):
                board_sum = 0
                for r in boards[i]:
                    board_sum += sum([x[0] for x in r if x[1] is not True])
                print(board_sum * n)
                exit()

File: 2021/4/part-one.ts
import { readFileSync } from "fs";
import { sum } from "../../utils";

async function main() {
  let result = 0;
  const file = readFileSync(process.argv[2], "utf-8");
  const rows = file
    .trim()
    .split("\n\n")
    .map((r) => {
      const row = r;
      return row;
    });

  const numbers = rows
    .splice(0, 1)[0]
    .split(",")
    .map((n) => parseInt(n.trim()));
  let boards = rows.map((r) =>
    r.split("\n").map((row) =>
      row
        .split(" ")
        .filter((num) => !isNaN(parseInt(num)))
        .map((num) => ({
          number: parseInt(num.trim()),
          marked: false,
        }))
    )
  );

  let boardsum = 0,
    finalNumber = 0;

  for (let n of numbers) {
    for (let i = 0; i < boards.length; i++) {
      boards[i] = boards[i].map((r) =>
        r.map((c) => {
          if (c.number === n) {
            return { number: c.number, marked: true };
          }
          return c;
        })
      );
      if (findBingo(boards[i])) {
        finalNumber = n;
        boardsum = sum(
          boards[i].flatMap((r) =>
            r.filter((c) => !c.marked).map((c) => c.number)
          )
        );
        break;
      }
    }
    if (finalNumber !== 0) break;
  }

  result = boardsum * finalNumber;
  console.log(result);
}

function findBingo(
  board: Array<{ number: number; marked: boolean }[]>
): boolean {
  let rows = board.map((r) => 0);
  let columns = board.map((r) => 0);
  if (board.length > 5) console.log(board);
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board.length; j++) {
      if (board[i][j].marked) {
        rows[i] += 1;
        columns[j] += 1;
      }
    }
  }
  if (rows.includes(board.length) || columns.includes(board.length)) {
    return true;
  }
}

main();
